Add dns-persist-01 challenge type to the manual plugin (#10633)
certbot/certbot: certbot/docs/using.rst

                      | one way to obtain wildcard certificates from Let's
                      | Encrypt.
manual_     Y    N    | Obtain a certificate by manually following instructions to    http-01_ (80),
                      | perform domain validation yourself. Certificates created this dns-01_ (53),
                      | way do not support autorenewal.                               or dns-persist-01_ (53)
                      | Autorenewal may be enabled by providing an authentication
                      | hook script to automate the domain validation steps. Using
                      | the ``dns-persist-01`` challenge type is another way to
                      | obtain wildcard certificates from Let's Encrypt.
=========== ==== ==== =============================================================== =============================

.. |dns_plugs| replace:: :ref:`DNS plugins <dns_plugins>`

Under the hood, plugins use one of several ACME protocol challenges_ to
prove you control a domain. The challenge options are http-01_ (which uses port 80),
and the two DNS-based challenges: dns-01_, or dns-persist-01_ (which both

> require configuration of a DNS server on port 53, though that's often not the

Reviewer: I wonder if people will be more likely to run dns-persist from a webserver. Like, since it doesn't need to hold onto the creds, just run it on the webserver since that'll be up and ready to cron. If that's the case, we may want to change the wording here a little. "that may not be the same machine" or something.

```suggestion
require configuration of a DNS server on port 53, though that may not be the
```

to copy and paste commands into another terminal session, which may
be on a different computer.

The manual plugin can use either the ``http``, ``dns``, or ``dns-persist``
challenges. You can use the ``--preferred-challenges`` option to choose the
challenge of your preference.

The ``http`` challenge will ask you to place a file with a specific name and
specific content in the ``/.well-known/acme-challenge/`` directory directly
in the top-level directory (“web root”) containing the files served by your
webserver. In essence it's the same as the webroot_ plugin, but not automated.

> When using the ``dns`` or ``dns-persist` challenges, ``certbot`` will ask you to

Reviewer: ```suggestion
When using the ``dns`` or ``dns-persist`` challenges, ``certbot`` will ask you to
```

place a TXT DNS record with specific contents under the domain name consisting
of the hostname for which you want a certificate issued, prepended by a
subdomain (either ``_acme-challenge`` or ``_validation-persist``, depending on
the challenge).

To perform a ``dns`` challenge with the manual plugin, you'd invoke Certbot like this:

::

        certbot certonly --preferred-challenges dns -d example.com --manual

The CA would then provide Certbot with a single-use validation token, which you (or
your auth hook script) would then use to create a TXT record at the subdomain
``_acme-challenge``, like this:

::

        _acme-challenge.example.com. 300 IN TXT "gfj9Xq...Rg85nM"

This validation token is only valid for this challenge, and once the CA
validates it, the TXT record can be removed.

The ``dns-persist`` challenge is different, and allows you to create a single
long-lived DNS record for certificate issuance and renewals. To perform a
``dns-persist`` challenge with the manual plugin, you'd invoke Certbot like
this:

::

        certbot certonly --preferred-challenges dns-persist -d example.com --manual

> Afterward, you (or your auth hook script) will be prompted to create a TXT record like this:

Reviewer: ```suggestion
Afterward, you (or your setup hook script) will be prompted to create a TXT record like this:
```

::

        _validation-persist.example.com. IN TXT "authority.example; accounturi=https://ca.example/acct/123"

This record can persist indefinitely, and as long as it's available, any future
certificate issuances at that subdomain will automatically be approved by the
CA. As such, ``dns-persist`` allows for easy automated certificate renewals.

> However, because Certbot's manual plugin requires manual intervention unless an
> auth hook script is provided, fully automated renewals with ``dns-persist``
> requires some sort of dummy auth hook, like this:
> 
> ::
> 
>         certbot certonly --preferred-challenges dns-persist-01 -d example.com --manual --manual-auth-hook=/bin/true

Reviewer: ```suggestion
When the ``dns-persist`` challenge is chosen, the ``--manual-setup-hook`` is not saved, and therefore will not be run again unless explicitly provided. The ``--manual-auth-hook`` is never used in ``dns-persist``.
```

The ``dns-persist`` challenge type also supports the issuance of wildcard
certificates. If the manual plugin detects that you're issuing a certificate for
a wildcard domain (e.g. ``*.example.com``), you (or your auth hook) will be
instructed to add ``policy=wildcard`` to the TXT record for the wildcard's
parent domain (e.g. ``example.com``).

.. _manual-renewal:

**Renewal with the manual plugin**

Certificates created using ``--manual`` only support automatic renewal with the
``dns-persist`` challenge, or when combined with an `authentication hook script
<#hooks>`_ via ``--manual-auth-hook`` to automatically set up the required HTTP
and/or TXT challenges.

> If your certificate was issued via the ``dns-persist`` challenge, automated renewal
> is possible by providing a dummy auth hook, for example:
> 
> ::
> 
>         certbot certonly --preferred-challenges dns-persist-01 -d example.com --manual --manual-auth-hook=/bin/true

Reviewer: ```suggestion
If your certificate was issued via the ``dns-persist`` challenge, automated renewal should happen in any case where automated renewal has otherwise been set up.
```
